internal/controller/clusterscmprovider_controller.go

```diff
@@ -69,3 +94,162 @@ func (r *ClusterScmProviderReconciler) SetupWithManager(ctx context.Context, mgr
 
 	return nil
 }
+
+func (r *ClusterScmProviderReconciler) handleFinalizer(ctx context.Context, clusterScmProvider *promoterv1alpha1.ClusterScmProvider) (bool, error) {
```

Comment:
At the risk of being too clever, consider pulling common finalizer handling logic into a util function. The function would:

1. Add the finalizer if needed
2. quit early if the object is being deleted and no finalizer is set
3. call cleanup logic (closing pull requests, checking for dangling dependencies, etc.)
4. if cleanup is done, remove the finalizer

Give it a try. If the logic is too messy, forget it.